!syscalls/threadsinfo: return more granular thread information

oI0ck opened the pull request:
## Description
This commit means to make `threadinfo` more flexible, by being able to query a thread identified by `tid` and query for specific thread attributes as specified in `flags` bitmask.

Depends-On: phoenix-rtos-utils#251
Depends-On: libphoenix#454

## Motivation and Context
`threadsinfo` is not very flexible and causes unnecessary overhead in programs using it.

Moreover, previously there was no was of inexpensively checking amount of present threads in the kernel, therefore the buffer passed to `threadsinfo` may have been often reallocated to check all threads. Now `threadsinfo`, if passed a `PH_THREADINFO_OPT_THREADCOUNT` flag will yield no information in passed buffer, but will return amount of currently present threads.

JIRA: RTOS-1187

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

## How Has This Been Tested?
- [ ] Already covered by automatic testing.
- [ ] New test added: (test PR is WIP).
- [X] Tested by hand on: `ia32-generic-qemu`, `riscv64-generic-qemu`, `sparcv8leon-generic-qemu`, `armv7a9-zynq7000-qemu`, `armv7r5f-zynqmp-qemu`.

## Checklist:
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [X] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- https://github.com/phoenix-rtos/libphoenix/pull/454
- https://github.com/phoenix-rtos/phoenix-rtos-utils/pull/251
- https://github.com/phoenix-rtos/phoenix-rtos-doc/pull/247
- [X] I will merge this PR by myself when appropriate.

adamgreloch commented on proc/threads.c:
```diff
@@ -2086,120 +2088,197 @@ void proc_threadsDump(u8 priority)
+
+
+/* call with threads_common.lock set */
+void proc_threadInfo(thread_t *thread, unsigned int flags, threadinfo_t *info)
```
If expected to be called with a lock, name it `_proc_threadInfo` (`_` implies need for external synchronization)

adamgreloch commented on proc/threads.c:
```diff
@@ -2086,120 +2088,197 @@ void proc_threadsDump(u8 priority)
+		if ((flags & PH_THREADINFO_OPT_THREADCOUNT) == PH_THREADINFO_OPT_THREADCOUNT) {
+			do {
+				i++;
+				t = lib_idtreeof(thread_t, idlinkage, lib_idtreeNext(&t->idlinkage.linkage));
```
I wonder if we shouldn't just maintain a global thread count instead

oI0ck replied:
Might be, though I'd say that's good enough for the amount of changes here. We'd have to guarantee that this counter will be kept in sync with node count in tree, which I'd say is besides the point in this PR.

adamgreloch replied:
Yeah, you're right. Maybe add a TODO/REVISIT then for now?

oI0ck commented on proc/threads.c:
```diff
@@ -2086,120 +2088,197 @@ void proc_threadsDump(u8 priority)
+		info->ppid = 0;
+	}
+
+	if ((flags & PH_THREADINFO_TID) == PH_THREADINFO_TID) {
```
The difference is syntactical, the compiler should emit the same code in both cases, I guess I just felt like writing it like that :D. I will change it though, for the sake of shorter line length.

oI0ck commented on proc/threads.c:
```diff
@@ -2086,120 +2088,197 @@ void proc_threadsDump(u8 priority)
-		t = lib_idtreeof(thread_t, idlinkage, lib_idtreeNext(&t->idlinkage.linkage));
+		proc_threadInfo(t, flags, info);
+		i++;
 	}
```
Right, that is pretty ugly. I've rewritten `PH_THREADINFO_THREADS_ALL` as a `do-while` loop and it's better.

oI0ck replied:
Yes, I know what you mean, but it still looks a bit better, since in `PH_THREADINFO_THREADS_ALL` case, we will have to use the loop either way, so making two loops inside of two branches (when they do mostly the same thing) is kinda ugly and the way I rewrote it also addresses having two return points.

adamgreloch replied:
There is still one return in `else`, that's what confused me, but if you think that forcibly reducing the divergence would clutter the code more than solve anything here, then it's fine by me

julianuziemblo reviewed (commented):
MISRA nits (for a better (worse) evening)

oI0ck commented on include/sysinfo.h:
```diff
@@ -17,6 +17,33 @@
+		PH_THREADINFO_VMEM | \
+		PH_THREADINFO_PPID)
+
+#define PH_THREADINFO_OPT_THREADCOUNT (1UL << 10U)
```
Done

julianuziemblo reviewed (commented):
More comments to `threadsinfo_t` data structure. Some of its members seem to have improper types with no explanation and we could fix it while at it.

adamgreloch commented on include/sysinfo.h:
Internally tid is an idtree key which is signed to support POSIX handles

adamgreloch replied:
Yes, I agree. My (vague) comment's intention was to explain why `tid` sometimes happens to be an `int` and AFAIK it's purely a consequence of idtree implementation.

We should probably clean this up, make it consistently `unsigned int` across the kernel or even introduce `tid_t` to control the types even further (outside of this PR).

oI0ck commented on proc/threads.c:
```diff
@@ -2086,120 +2088,189 @@ void proc_threadsDump(u8 priority)
+		entry = thread->process->entries;
+		if (entry != NULL) {
+			do {
+				vmem += (int)entry->size;
```
I'm dropping the `vmem` type change, correcting those type may have bigger implications and should be considered on its own.

oI0ck commented on proc/threads.c:
```diff
@@ -1874,95 +1875,165 @@
+	else {
+		t = lib_idtreeof(thread_t, idlinkage, lib_idtreeFind(&threads_common.id, tid));
+		if (t == NULL) {
+			(void)proc_lockClear(&threads_common.lock);
```
Unrelated to this PR

julianuziemblo commented on include/sysinfo.h:
```diff
@@ -20,15 +20,15 @@
+#define PH_THREADINFO_WAITING ((unsigned int)1UL << 6)
+#define PH_THREADINFO_NAME    ((unsigned int)1UL << 7)
+#define PH_THREADINFO_VMEM    ((unsigned int)1UL << 8)
+#define PH_THREADINFO_PPID    ((unsigned int)1UL << 9)
```
`PS-C-REC-058`: when shifting `<=7` bits, mark the `1` with `U` instead of `UL`. This eliminates the need for casting, but looks inconsistent. It's only a recommendation and not a rule though.

oI0ck commented on proc/threads.c:
```diff
@@ -1874,95 +1875,165 @@
+	if ((flags & PH_THREADINFO_PRIO) != 0) {
+		info->priority = (int)thread->priorityBase;
+	}
+
```
Corrected

oI0ck commented on proc/threads.c:
```diff
@@ -1874,95 +1875,165 @@
-			hal_memcpy(tinfo.name, "[idle]", sizeof("[idle]"));
+			hal_memcpy(info->name, "[idle]", sizeof("[idle]"));
 		}
+	}
```
Corrected

julianuziemblo commented on proc/threads.h:
```diff
@@ -117,13 +117,13 @@ int proc_join(int tid, time_t timeout);
 
 
-typedef void (*proc_threadsListCb_t)(void *arg, int i, threadinfo_t *info);
+typedef void (*proc_threadsListCb_t)(void *arg, threadinfo_t *info);
```
ik this is not a change from this PR, but `PS-C-REC-015` states that function types should be suffixed with `Fn_t`

julianuziemblo commented on proc/threads.c:
```diff
@@ -1881,95 +1882,165 @@ void proc_threadsDump(u8 priority)
+	else {
+		info->pid = 0;
+		info->ppid = 0;
+	}
```
there was a TODO here for `ppid`:

> /* TODO: tinfo.ppid = t->process->parent != NULL ? t->process->parent->id : 0; */

I'd say it should still be here as this wasn't fixed

julianuziemblo reviewed (dismissed):
nits, otherwise LGTM

adamgreloch commented on proc/threads.c:
```diff
@@ -2086,120 +2088,197 @@ void proc_threadsDump(u8 priority)
+}
+
+
+static inline int _proc_calculateVmem(thread_t *thread)
```
Please add a comment similar to one preceeding `_proc_threadInfo`

adamgreloch replied:
Bump

oI0ck commented on proc/threads.h:
```diff
@@ -117,13 +117,13 @@ int proc_join(int tid, time_t timeout);
 
 
-typedef void (*proc_threadsListCb_t)(void *arg, int i, threadinfo_t *info);
+typedef void (*proc_threadsListCb_t)(void *arg, threadinfo_t *info);
```
done

oI0ck commented on proc/threads.c:
```diff
@@ -1881,95 +1882,165 @@ void proc_threadsDump(u8 priority)
+	now = _proc_gettimeRaw();
+	if (thread->process != NULL) {
+		info->pid = process_getPid(thread->process);
+		info->ppid = 0;
```
done

etiaro commented on proc/threads.c:
```diff
@@ -1907,97 +1908,165 @@ void proc_threadsDump(u8 priority)
+
+#ifdef NOMMU
+	if (thread->process != NULL) {
+		entry = thread->process->entries;
```
I see that this is also the case on master, but process->entries are not synchronized here in any way (map.c takes process->lock and map-> lock for every operation on these).

oI0ck replied:
We don't access the `vm_map_t` here so `process->lock` may do here?

oI0ck commented:
Whoops, closed by mistake. TIL that github will automatically close the PR if it has nothing to merge (makes sense), I'll open a new one.

There is a problem with this PR on zynq7000-zedboard that I can't seem to figure out, seemingly commenting out all of this PR's content will still cause zynq to hang during tests. Will reopen when I get rid of those issues.

etiaro commented:
You should be able to reopen it after pushing some commits on this branch

oI0ck committed:
!syscalls/threadsinfo: return more granular thread information

threadsinfo is not very flexible, takes a long time and causes
unnecessary overhead in programs using it.

This commit means to make it more flexible, by being able to query
a thread identified by tid and query for specific thread attributes
as specified in flags bitmask.

Moreover, previously there was no way of inexpensively checking amount
of present threads in the kernel, therefore the buffer passed to
threadinfo may have been often reallocated to check all threads.
Now threadsinfo, if passed a PH_THREADINFO_OPT_THREADCOUNT flag
will yield no information in passed buffer, but will return amount of
currently present threads.

TASK: RTOS-1187

etiaro commented:
NIT: Typo in commit message:
`Moreover, previously there was no was`

ziemleszcz commented:
Please remove the jmaksymowicz commit from this PR.

ziemleszcz commented on proc/threads.c:
```diff
@@ -1907,97 +1908,164 @@ void proc_threadsDump(u8 priority)
+			i++;
+		} while (t != NULL);
+	}
+	else {
```
Missing `n >= 1` check. Note that `vm_mapBelongs()` check passes on `n==0`.

adamgreloch replied:
This seems unresolved too

oI0ck replied:
The check was added in `syscalls.c`

oI0ck replied:
Right, added the check here, 0 should be allowed for getting the thread count

ziemleszcz commented on proc/threads.c:
```diff
@@ -1907,97 +1908,164 @@ void proc_threadsDump(u8 priority)
+
+	hal_spinlockSet(&threads_common.spinlock, &sc);
+	now = _proc_gettimeRaw();
+	hal_spinlockClear(&threads_common.spinlock, &sc);
```
Other fields should also be read under spinlock - check old `procThreadIter()`.

ziemleszcz commented on proc/threads.c:
```diff
@@ -1907,97 +1908,164 @@ void proc_threadsDump(u8 priority)
-		cb(arg, i, &tinfo);
+			t = lib_idtreeof(thread_t, idlinkage, lib_idtreeNext(&t->idlinkage.linkage));
+			i++;
+		} while (t != NULL);
```
Change to `while(t != NULL) {}`.

ziemleszcz commented on proc/threads.c:
```diff
@@ -2006,11 +2074,37 @@ int proc_threadsIter(int n, proc_threadsListCb_t cb, void *arg)
+		}
+
+		t = lib_idtreeof(thread_t, idlinkage, lib_idtreeNext(&t->idlinkage.linkage));
+	} while (t != NULL);
```
Change to while(t != NULL) {}.

oI0ck commented on proc/threads.c:
```diff
@@ -1920,97 +1921,166 @@
+			} while (entry != thread->process->entries);
+		}
+	}
+	proc_lockClear(&thread->process->lock);
```
resolved

oI0ck commented on proc/threads.c:
```diff
@@ -1920,97 +1921,166 @@
 	map_entry_t *entry;
+
+#ifdef NOMMU
+	proc_lockSet(&thread->process->lock);
```
resolved

ziemleszcz commented on proc/threads.c:
```diff
@@ -1937,97 +1938,171 @@ void proc_threadsDump(u8 priority)
+	}
+	else {
+		if (n < 1) {
+			return -EINVAL;
```
Return with lock taken.

ziemleszcz commented on proc/threads.c:
```diff
@@ -1937,97 +1938,171 @@ void proc_threadsDump(u8 priority)
-				} while (entry != t->process->entries);
-			}
+	if ((flags & PH_THREADINFO_PPID) != 0U) {
+		ppid = posix_getppid(info->pid);
```
`posix_gettpid()` takes `posix_common.lock`. Since this entire function runs with `threads_common.lock` already taken, we end up taking two global locks. This should be avoided - see RTOS-1095 on YT. The original implementation took these two locks separately.

ziemleszcz replied:
Calling `posix_gettpid()` twice now.

ziemleszcz commented on proc/threads.c:
```diff
@@ -1937,110 +1938,225 @@ void proc_threadsDump(u8 priority)
 	(void)proc_lockClear(&threads_common.lock);
 
+	if ((flags & PH_THREADINFO_PPID) != 0U) {
+		for (int i = 0; i < n; i++) {
```
Shadowing first `i` variable (return count). Declaring variable in this place is forbidded in the kernel.
